sound/soc/sof/intel/cnl.c

```diff
@@ -171,23 +171,53 @@ static bool cnl_compact_ipc_compress(struct snd_sof_ipc_msg *msg,
 static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
 			    struct snd_sof_ipc_msg *msg)
 {
+	struct sof_intel_hda_dev *hdev = sdev->pdata->hw_pdata;
+	struct sof_ipc_cmd_hdr *hdr;
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_D0,
+		.substate = SOF_HDA_DSP_D0I0,
+	};
 	u32 dr = 0;
 	u32 dd = 0;
+	int ret;
 
 	if (cnl_compact_ipc_compress(msg, &dr, &dd)) {
 		/* send the message via IPC registers */
 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDD,
 				  dd);
 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
 				  CNL_DSP_REG_HIPCIDR_BUSY | dr);
-	} else {
-		/* send the message via mailbox */
-		sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
-				  msg->msg_size);
-		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
-				  CNL_DSP_REG_HIPCIDR_BUSY);
+		return 0;
+	}
+
+	/* ensure the DSP is in D0I0 before sending the message via mailbox */
+	ret = snd_sof_dsp_set_power_state(sdev, target_state);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: resuming DSP %d\n", ret);
+		return ret;
 	}
 
+	/* send the message via mailbox */
+	sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
+			  msg->msg_size);
+	snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
+			  CNL_DSP_REG_HIPCIDR_BUSY);
+
+	hdr = msg->msg_data;
+
+	/*
+	 * Do not schedule delayed work for DSP D0i3 entry
+	 * when sending CTX_SAVE IPC.
```

Comment:
this is unclear to me, you only use the mod_delayed_work when you are NOT handling CTX_SAVE?